### PR TITLE
Feature/rdsssam 68 2.1.0 crud conformance

### PR DESCRIPTION
**Pull request by howard-jisc**

### CODE REVIEW

File: willow/lib/rdss/actors/append_to_numbered_hash.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module Rdss
+  module Actors
+    class AppendToNumberedHash < SetHashValues
+
+      def to_numbered_hash(arr)
```
Comment (tripledub) [inline willow/lib/rdss/actors/append_to_numbered_hash.rb]: You could have done this in a 1-liner with
```ruby
Hash[(0...arr.size).zip arr]
```
(untested 🤔 )

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb [hunk 2]
```diff
@@ -19,47 +34,91 @@
...
     let(:attributes) { {:object_version => "1", :title => ["another test title"]} }
-    let(:rdss_cdm) { create(:rdss_cdm, title: ["test title"], object_version: "1", state: inactive_state) }
-    let(:env) { Hyrax::Actors::Environment.new(rdss_cdm, ability, attributes) }
+    let(:env) { Hyrax::Actors::Environment.new(unapproved_rdss_cdm, ability, attributes) }
+
     it 'object version increments to 2' do
       expect { middleware.update(env) }.not_to change { env.attributes[:object_version] }
     end
+
+    it "object related identifiers change to not include previous object_uuid as is_new_version_of" do
```
Comment (tripledub) [inline willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: Just a nit-pick but prefer single quotes where there will be no string interpolation

**Comment by tripledub** [comment]
Other than comment @blackrat, mine are just nit-picks so it is a 👍 from me

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb [hunk 3]
```diff
@@ -4,9 +4,24 @@
...
+      title: ["test title"], 
+      object_version: "1", 
+      object_uuid: original_object_uuid, 
+      state: active_state) 
+    }
+
+  let(:unapproved_rdss_cdm) { 
+      create(:rdss_cdm, 
+        title: ["test title"], 
+        object_version: "1",
```
Comment (blackrat) [inline willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: Is this create going through the actor when FactoryGirl creates the object?

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb [hunk 4]
```diff
@@ -19,47 +34,91 @@
...
+    end
   end
 
   describe "minor update" do
-    let(:attributes) { {:object_version => "1", :title => ["test title"]} }
-    let(:rdss_cdm) { create(:rdss_cdm, title: ["test title"], object_version: "1", state: active_state) }
-    let(:env) { Hyrax::Actors::Environment.new(rdss_cdm, ability, attributes) }
-    it 'object version remains 1' do
-      expect { middleware.update(env) }.not_to change { env.attributes[:object_version] }
+    let(:attributes) { {:object_version => "1", :title => ["test title"]} }
```
Comment (blackrat) [inline willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: object_version shouldn't exist in here unless it's being set in the form. That may be a different test to run.

File: willow/app/actors/hyrax/actors/rdss_cdm_object_versioning_actor.rb [hunk 5]
```diff
@@ -1,13 +1,22 @@
...
       def create(env)
-        env.attributes[:object_version]='1' if env.attributes[:object_version].blank?
+        ::Rdss::Actors::SetAttributeValuesIfBlank.(env, create_default_values)
         next_actor.create(env)
       end
-
+      
       def update(env)
-        env.attributes[:object_version].next! if CurationConcernApproved.(env) && SignificantFieldsChanged.(env) unless ObjectVersionChanged.(env)
+        ::Rdss::Actors::PerformCdmVersioning.(env) if CurationConcernApproved.(env)
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/app/actors/hyrax/actors/rdss_cdm_object_versioning_actor.rb]: added the guard back in

File: willow/lib/rdss/actors/append_to_numbered_hash.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/lib/rdss/actors/append_to_numbered_hash.rb]: marked it private

File: willow/lib/rdss/actors/append_to_numbered_hash.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        num_hash = {}
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/lib/rdss/actors/append_to_numbered_hash.rb]: similar to Stew's comment i think - i've left it for now in the interests of moving forward.

File: willow/lib/rdss/actors/append_to_numbered_hash.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        arr.each_with_index {|v, i| num_hash[i.to_s] = v}
+        num_hash
+      end
+      
+      public
+      def call(hash)
+        hash.each { |key, value|
+          current_values = attributes[key].nil? ? [] : attributes[key].values
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/lib/rdss/actors/append_to_numbered_hash.rb]: left as is, as discussed.

File: willow/lib/rdss/actors/append_to_numbered_hash.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+module Rdss
+  module Actors
+    class AppendToNumberedHash < SetHashValues
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/lib/rdss/actors/append_to_numbered_hash.rb]: left unchanged, as discussed.

File: willow/lib/rdss/actors/perform_cdm_versioning.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+            _destroy: 'false', 
+            relation_type: 'is_new_version_of', 
+            identifier_attributes: {
+              identifier_value: uuid, 
+              identifier_type: 'source_id'
+            }
+          }
+        end
+
+        def record_current_uuid(env)
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/lib/rdss/actors/perform_cdm_versioning.rb]: yes

File: willow/spec/actors/hyrax/actors/rdss_cdm_actor_spec.rb [hunk 10]
```diff
@@ -8,7 +8,7 @@
   let(:env) { Hyrax::Actors::Environment.new(rdss_cdm, ability, attributes) }
   let(:terminator) { Hyrax::Actors::Terminator.new }
   let(:depositor) { create(:user) }
-  let(:attributes) { {:title => ["a test title"]} }
+  let(:attributes) { {:title => ["a test title"], object_value: nil} }
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/spec/actors/hyrax/actors/rdss_cdm_actor_spec.rb]: made changes, as discussed

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb [hunk 11]
```diff
@@ -19,47 +34,91 @@
     let(:attributes) { {:object_version => "", :title => ["test title"]} }
```
[the remark this answers is not in the document]
Comment (howard-jisc) [reply willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: removed the use of :object_version in attributes, as discussed

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb (hunk 3 above)
Comment (howard-jisc) [reply willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: unchanged, as discussed

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb (hunk 4 above)
Comment (howard-jisc) [reply willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: yes

File: willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb (hunk 2 above)
Comment (howard-jisc) [reply willow/spec/actors/hyrax/actors/rdss_cdm_object_versioning_actor_spec.rb]: made the changes

File: willow/lib/rdss/actors/append_to_numbered_hash.rb (hunk 1 above)
Comment (howard-jisc) [reply willow/lib/rdss/actors/append_to_numbered_hash.rb]: left as is for now

**Pull request merged**